File: feature/src/main/java/org/indiv/dls/games/vocabrecall/feature/PuzzleFragment.java
package org.indiv.dls.games.vocabrecall.feature;

import java.util.ArrayList;
import java.util.List;

import org.indiv.dls.games.vocabrecall.feature.db.GameWord;


import android.content.Context;
import android.content.res.Resources;
import android.graphics.Color;
import android.os.Bundle;
import android.os.Vibrator;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentActivity;
import android.widget.Space;
import android.util.TypedValue;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.View.OnClickListener;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

public class PuzzleFragment extends Fragment {

    // interface for activity to implement to receive touch event
    public interface PuzzleListener {
        void onPuzzleClick(GameWord gameWord);
    }

    private static final String TAG = PuzzleFragment.class.getSimpleName();
    private static final int TENTATIVE_COLOR = 0xFFAAAAAA; // Color.LTGRAY is 0xFFCCCCCC, Color.GRAY is 0xFF888888
    private static final int CONFIDENT_COLOR = Color.BLACK; // Color.LTGRAY is 0xFFCCCCCC, Color.GRAY is 0xFF888888
    private static final int CELL_BKGD_LEVEL_NORMAL = 1;
    private static final int CELL_BKGD_LEVEL_ERRORED = 2;
    private static final int CELL_BKGD_LEVEL_SELECTED = 3;
    private static final int CELL_BKGD_LEVEL_ERRORED_SELECTED = 4;

    private TableLayout mTableLayout;
    private GridCell[][] mCellGrid;
    private int mGridWidth;
    private int mGridHeight;
    private int mPixelsPerCell;
    private GameWord mCurrentGameWord;
    private Vibrator mVibrator;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {

        // inflate the view
        View view = inflater.inflate(R.layout.fragment_puzzle, container);

        // Get instance of Vibrator from current Context
        mVibrator = (Vibrator) getActivity().getSystemService(Context.VIBRATOR_SERVICE);


//		Resources r = getResources();
//		DisplayMetrics displayMetrics = r.getDisplayMetrics();
//		
////		Configuration config = r.getConfiguration();
//		
////		int viewWidthPixels = displayMetrics.widthPixels;
////		int viewHeightPixels = displayMetrics.heightPixels - actionBarHeightInPixels;
//		int viewWidthPixels = view.getWidth();
//		int viewHeightPixels = view.getHeight();
//		
//
//		
//		mTableLayout = (TableLayout)view.findViewById(R.id.cellTable);
//		
//		
//		// calculate number of pixels equivalent to 24dp (24dp allows 13 cells on smallest screen supported by Android (320dp width, 426dp height))
//	    mPixelsPerCell = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 24, displayMetrics));
//		
//		
//		// set up grid
//		mGridHeight = (int)(viewHeightPixels / mPixelsPerCell) - 1;  // subtract 1 because subtracting action bar height doesn't seem to be enough
//		mGridWidth = viewWidthPixels / mPixelsPerCell - 1;  // subtract 1 to give margin for consistency with height
//		mCellGrid = new GridCell[mGridHeight][mGridWidth];
//
//		// create table rows 
//		for (int row = 0;  row < mGridHeight;  row++) {
//			TableRow tableRow = new TableRow(getActivity());
//			tableRow.setGravity(Gravity.CENTER);
//			mTableLayout.addView(tableRow);
//		}


        // Note that database not set up yet at this point (happening in other thread).
        // When it completes, it will call onFinishDbSetup().


        return view;
    }


    //---------------------------------------------------//
    //------------ end of overridden methods ------------//
    //---------------------------------------------------//

    public void initialize(int viewWidthPixels, int viewHeightPixels) {
        Resources r = getResources();
//		DisplayMetrics displayMetrics = r.getDisplayMetrics();

//		View view = getView();
//		Configuration config = r.getConfiguration();
//		int viewWidthPixels = displayMetrics.widthPixels;
//		int viewHeightPixels = displayMetrics.heightPixels - actionBarHeightInPixels;
//		int viewWidthPixels = view.getWidth();
//		int viewHeightPixels = view.getHeight();


        // calculate number of pixels equivalent to 24dp (24dp allows 13 cells on smallest screen supported by Android (320dp width, 426dp height))
//	    mPixelsPerCell = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, r.getDimension(R.dimen.cell_width), displayMetrics));
        mPixelsPerCell = Math.round(r.getDimension(R.dimen.cell_width));

        // set up grid
        mGridHeight = (int) (viewHeightPixels / mPixelsPerCell) - 2;  // subtract 2 because subtracting action bar height doesn't seem to be enough
        mGridWidth = viewWidthPixels / mPixelsPerCell - 1;  // subtract 1 to give margin for consistency with height
        mCellGrid = new GridCell[mGridHeight][mGridWidth];

        // create table rows
        mTableLayout = (TableLayout) getView().findViewById(R.id.cellTable);
        for (int row = 0; row < mGridHeight; row++) {
            TableRow tableRow = new TableRow(getActivity());
            tableRow.setGravity(Gravity.CENTER);
            mTableLayout.addView(tableRow);
        }
    }

    public boolean doWordsFitInGrid(List<GameWord> gameWords) {
        for (GameWord gw : gameWords) {
            if (gw.getRow() >= mGridHeight || gw.getCol() >= mGridWidth) {
                return false;
            }
            if (gw.isAcross() && gw.getCol() + gw.getWord().length() > mGridWidth) {
                return false;
            }
            if (!gw.isAcross() && gw.getRow() + gw.getWord().length() > mGridHeight) {
                return false;
            }
        }
        return true;
    }

    public GridCell[][] getCellGrid() {
        return mCellGrid;
    }


    /**
     * create new game (called first time app run, or when user starts new game)
     * implements interface for receiving callback from ConfirmStartNewGameDialogFragment
     */
    public void clearExistingGame() {
        // clear out any existing data
        mCurrentGameWord = null;
        for (int row = 0; row < mGridHeight; row++) {
            TableRow tableRow = (TableRow) mTableLayout.getChildAt(row);
            tableRow.removeAllViews();
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                if (gridCell != null) {
                    mCellGrid[row][col] = null;  // remove cell data
                }
            }
        }
    }


    /*
     */
    public GameWord getCurrentGameWord() {
        return mCurrentGameWord;
    }


    /*
     * implements interface for receiving callback from AnswerFragment
     */
    public List<TextView> getPuzzleRepresentation() {
        // show puzzle representation in layout
        Resources r = getResources();
        int wordLength = mCurrentGameWord.getWord().length();
        int row = mCurrentGameWord.getRow();
        int col = mCurrentGameWord.getCol();
//        int size = Math.round(TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, r.getDimension(R.dimen.cell_representation_width), r.getDisplayMetrics()));
        int size = Math.round(r.getDimension(R.dimen.cell_representation_width));
        float fontHeight = size * .75f;
        List<TextView> views = new ArrayList<TextView>();
        for (int charIndex = 0; charIndex < wordLength; charIndex++) {
            TextView textView = createCellTextView(r, fontHeight, size);
            views.add(textView);
            GridCell gridCell = mCellGrid[row][col];
            if (mCurrentGameWord.isAcross()) {
                if (gridCell.getGameWordDown() != null) {
                    fillTextView(textView, gridCell.getUserCharDown(), gridCell.getGameWordDown().isConfident());
                }
                col++;
            } else {
                if (gridCell.getGameWordAcross() != null) {
                    fillTextView(textView, gridCell.getUserCharAcross(), gridCell.getGameWordAcross().isConfident());
                }
                row++;
            }
        }
        return views;
    }


    public void createGrid() {
        Resources r = getResources();
        float fontHeight = mPixelsPerCell * .75f;

        // create click listener for puzzle clicks
        OnClickListener onPuzzleClickListener = new OnClickListener() {
            @Override
            public void onClick(View v) {
                GridCell gridCell = getCellForView(v);
                if (gridCell != null) {
                    if (mVibrator != null) {
                        mVibrator.vibrate(25); // vibrate for 300 milliseconds
                    }
                    setCurrentGameWord((gridCell.getGameWordDown() != null) ? gridCell.getGameWordDown() : gridCell.getGameWordAcross());
                    ((PuzzleListener) getActivity()).onPuzzleClick(mCurrentGameWord);
                }
            }
        };

        FragmentActivity activity = getActivity();

        // add views into table rows and columns
        GameWord firstGameWord = null;
        for (int row = 0; row < mGridHeight; row++) {
            TableRow tableRow = (TableRow) mTableLayout.getChildAt(row);
            tableRow.removeAllViews();
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                if (gridCell != null) {

                    // create text view for this row and column
                    TextView textView = createCellTextView(r, fontHeight, mPixelsPerCell);
                    textView.setOnClickListener(onPuzzleClickListener);
                    tableRow.addView(textView, col);
                    gridCell.setView(textView);

                    fillTextView(gridCell);

                    // set current game word to the first across word found for sake of dual pane mode
                    if (firstGameWord == null) {
                        firstGameWord = gridCell.getGameWordAcross();
                        if (firstGameWord == null) {
                            firstGameWord = gridCell.getGameWordDown();
                        }
                    }
                } else {
                    tableRow.addView(new Space(activity), col);
                }
            }
        }

        // do this after grid completely created
        if (firstGameWord != null) {
            setCurrentGameWord(firstGameWord);
        }
    }

    private void setCurrentGameWord(GameWord gameWord) {
        if (mCurrentGameWord != null) {
            showAsSelected(mCurrentGameWord, false);
        }
        mCurrentGameWord = gameWord;
        showAsSelected(mCurrentGameWord, true);
    }

    public boolean selectNextErroredGameWord() {
        for (int row = 0; row < mGridHeight; row++) {
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                if (gridCell != null && gridCell.hasUserError()) {
                    setCurrentGameWord(gridCell.getGameWordAcross() != null ? gridCell.getGameWordAcross() : gridCell.getGameWordDown());
                    return true;
                }
            }
        }
        return false;
    }

    public boolean showErrors(boolean showErrors) {
        int numErrors = 0;

        // update background of cells based on whether text is correct or not
        for (int row = 0; row < mGridHeight; row++) {
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                // if cell is part of currently selected game word, adjust the level for the background
                if (gridCell != null) {
                    boolean isSelected = (mCurrentGameWord.equals(gridCell.getGameWordAcross()) || mCurrentGameWord.equals(gridCell.getGameWordDown()));
                    TextView textView = gridCell.getView();
                    if (showErrors && gridCell.hasUserError()) {
                        numErrors++;
                        textView.getBackground().setLevel(isSelected ? CELL_BKGD_LEVEL_ERRORED_SELECTED : CELL_BKGD_LEVEL_ERRORED); // set error background
                        textView.setTextColor(Color.RED);
                    } else {
                        textView.getBackground().setLevel(isSelected ? CELL_BKGD_LEVEL_SELECTED : CELL_BKGD_LEVEL_NORMAL); // set normal text cell background
                        textView.setTextColor(gridCell.isDominantCharConfident() ? CONFIDENT_COLOR : TENTATIVE_COLOR);
                    }
                }
            }
        }
        return (numErrors > 0);
    }

    //
    // returns true if puzzle is completely filled in
    // (if correctly=true, then only returns true if all entries are filled in and correct)
    //
    public boolean isPuzzleComplete(boolean correctly) {
        for (int row = 0; row < mGridHeight; row++) {
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                if (gridCell != null) {
                    // if cell is empty, then not complete
                    if (gridCell.getDominantUserChar() == 0 || (correctly && gridCell.hasUserError())) {
                        return false;
                    }
                }
            }
        }
        return true;
    }


    /*
     *
     */
    public void updateUserTextInPuzzle(GameWord gameWord) {
        // show answer in puzzle
        String userText = gameWord.getUserText();
        int userTextLength = userText.length();
        int wordLength = mCurrentGameWord.getWord().length();
        int row = gameWord.getRow();
        int col = gameWord.getCol();
        for (int charIndex = 0; charIndex < wordLength; charIndex++) {
            GridCell gridCell = mCellGrid[row][col];
            char userChar = (charIndex < userTextLength) ? userText.charAt(charIndex) : 0;
            if (gameWord.isAcross()) {
                gridCell.setUserCharAcross(userChar);
                fillTextView(gridCell);
                col++;
            } else {
                gridCell.setUserCharDown(userChar);
                fillTextView(gridCell);
                row++;
            }
        }
    }

    //-----------------------------------------//
    //------------ private methods ------------//
    //-----------------------------------------//

    private void showAsSelected(GameWord gameWord, boolean asSelected) {
        if (gameWord != null) {

//			int animationId = gameWord.isAcross()? R.anim.word_selection_horiz : R.anim.word_selection_vert;
//			Animation selectionAnimation = asSelected? AnimationUtils.loadAnimation(getActivity(), animationId) : null;

            int row = gameWord.getRow();
            int col = gameWord.getCol();
            for (int i = 0; i < gameWord.getWord().length(); i++) {
                GridCell gridCell = mCellGrid[row][col];
                TextView textView = gridCell.getView();
                int newLevel = CELL_BKGD_LEVEL_NORMAL;
                switch (textView.getBackground().getLevel()) {
                    case CELL_BKGD_LEVEL_NORMAL:
                    case CELL_BKGD_LEVEL_SELECTED:
                        newLevel = asSelected ? CELL_BKGD_LEVEL_SELECTED : CELL_BKGD_LEVEL_NORMAL;
                        break;
                    case CELL_BKGD_LEVEL_ERRORED:
                    case CELL_BKGD_LEVEL_ERRORED_SELECTED:
                        newLevel = asSelected ? CELL_BKGD_LEVEL_ERRORED_SELECTED : CELL_BKGD_LEVEL_ERRORED;
                        break;
                }
                textView.getBackground().setLevel(newLevel);

//				if (selectionAnimation != null) {
//					textView.startAnimation(selectionAnimation);
//				}

                if (gameWord.isAcross()) {
                    col++;
                } else {
                    row++;
                }
            }
        }
    }

    private void fillTextView(GridCell gridCell) {
        fillTextView(gridCell.getView(), gridCell.getDominantUserChar(), gridCell.isDominantCharConfident());
    }

    private void fillTextView(TextView textView, char dominantUserChar, boolean confident) {
        if (dominantUserChar != 0) {
            textView.setTextColor(confident ? CONFIDENT_COLOR : TENTATIVE_COLOR);
            textView.setText(String.valueOf(dominantUserChar));
        } else {
            textView.setText(null);
        }
    }

    private TextView createCellTextView(Resources r, float fontHeight, int size) {
        TextView textView = new TextView(getActivity());
        textView.setGravity(Gravity.CENTER);
        // need to create Drawable object for each TextView
        textView.setBackgroundDrawable(r.getDrawable(R.drawable.cell_drawable));  // using deprecated method since setBackground() not supported until API level 16
        textView.setTextSize(TypedValue.COMPLEX_UNIT_PX, fontHeight);
        textView.setWidth(size);
        textView.setHeight(size);
        textView.getBackground().setLevel(1); // set normal text cell background (i.e. no error indication)
//		textView.setSoundEffectsEnabled(false); // true by default, consider disabling since we're providing our own vibration (except not all devices have vibration)
        return textView;
    }


    private GridCell getCellForView(View v) {
        for (int row = 0; row < mGridHeight; row++) {
            for (int col = 0; col < mGridWidth; col++) {
                GridCell gridCell = mCellGrid[row][col];
                if (gridCell != null && v == gridCell.getView()) {
                    return gridCell;
                }
            }
        }
        return null;
    }

}

File: feature/src/main/java/org/indiv/dls/games/verboscruzados/feature/PuzzleFragment.kt
package org.indiv.dls.games.verboscruzados.feature


import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.ScrollView
import android.widget.TableRow
import kotlinx.android.synthetic.main.fragment_puzzle.*
import org.indiv.dls.games.verboscruzados.feature.component.PuzzleCellTextView
import org.indiv.dls.games.verboscruzados.feature.game.GameWord


/**
 * Fragment containing the crossword puzzle.
 */
class PuzzleFragment : Fragment() {

    //region PUBLIC INTERFACES ---------------------------------------------------------------------

    // interface for activity to implement to receive touch event
    interface PuzzleListener {
        fun onPuzzleClick(gameWord: GameWord)
    }

    //endregion

    //region PRIVATE PROPERTIES --------------------------------------------------------------------

    private var gridWidth: Int = 0
    private var gridHeight: Int = 0
    private lateinit var vibration: Vibration

    //endregion

    //region PUBLIC PROPERTIES ---------------------------------------------------------------------

    var initialized = false
    lateinit var cellGrid: Array<Array<GridCell?>>

    var selectedCellIndex = 0

    var currentGameWord: GameWord? = null
        private set(gameWord) {
            field?.let { showAsSelected(it, false) }   // unselect previous word
            gameWord?.let { showAsSelected(it, true) } // select new word
            field = gameWord
        }

    var scrollPosition: Int
        get() = (view as? ScrollView)?.scrollY ?: 0
        set(value) {
            (view as? ScrollView)?.smoothScrollTo(0, value)
        }

    //endregion

    //region OVERRIDDEN FUNCTIONS ------------------------------------------------------------------

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_puzzle, container)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // Get instance of Vibrator from current Context
        context?.let {
            vibration = Vibration(it)
        }
    }

    //endregion

    //region PUBLIC FUNCTIONS ----------------------------------------------------------------------

    fun initialize(gridWidth: Int, gridHeight: Int) {
        this.gridWidth = gridWidth
        this.gridHeight = gridHeight

        // set up grid
        cellGrid = Array(gridHeight) { arrayOfNulls<GridCell>(gridWidth) }

        // create table rows
        for (row in 0 until gridHeight) {
            val tableRow = TableRow(activity)
            tableRow.gravity = Gravity.CENTER
            cell_table_layout.addView(tableRow)
        }

        initialized = true
    }

    fun doWordsFitInGrid(gameWords: List<GameWord>): Boolean {
        gameWords.forEach {
            if ((it.row >= gridHeight || it.col >= gridWidth) ||
                    (it.isAcross && it.col + it.word.length > gridWidth) ||
                    (!it.isAcross && it.row + it.word.length > gridHeight)) {
                return false
            }
        }
        return true
    }

    fun clearExistingGame() {
        // clear out any existing data
        currentGameWord = null
        for (row in 0 until gridHeight) {
            val tableRow = cell_table_layout.getChildAt(row) as TableRow
            tableRow.removeAllViews()
            cellGrid[row].fill(null)
        }
    }

    /**
     * Creates grid of textviews making up the puzzle.
     *
     * @param puzzleListener implementation of the [PuzzleListener] interface that listens for clicks on the puzzle.
     */
    fun createGrid(puzzleListener: PuzzleListener) {
        val onPuzzleClickListener = OnClickListener { v ->
            getCellForView(v)?.let {gridCell ->
                vibration.vibrate()

                val sameWordSelected = currentGameWord == gridCell.gameWordDown || currentGameWord == gridCell.gameWordAcross
                val newGameWord = if (sameWordSelected) {
                    currentGameWord
                } else {
                    gridCell.gameWordDown ?: gridCell.gameWordAcross
                }

                newGameWord?.let {
                    selectedCellIndex = if (it.isAcross) gridCell.acrossIndex else gridCell.downIndex
                    currentGameWord = newGameWord  // this assignment will call setter which will take care of showing word and cell as selected

                    // Notify the listener regardless of whether new word selected or not (e.g. so keyboard can be shown).
                    puzzleListener.onPuzzleClick(it)
                }
            }
        }

        // add views into table rows and columns
        var firstGameWord: GameWord? = null
        for (row in 0 until gridHeight) {
            val tableRow = cell_table_layout.getChildAt(row) as TableRow
            tableRow.removeAllViews()
            for (col in 0 until gridWidth) {
                cellGrid[row][col]?.let {
                    // create text view for this row and column
                    val textView = PuzzleCellTextView(context!!)
                    textView.setOnClickListener(onPuzzleClickListener)
                    tableRow.addView(textView, col)
                    it.view = textView

                    fillTextView(it)

                    // set current game word to the first across word found
                    if (firstGameWord == null) {
                        firstGameWord = it.gameWordAcross ?: it.gameWordDown
                    }
                }
                        ?: run {
                            tableRow.addView(View(activity), col)
                        }
            }
        }

        // make the initial word selection
        if (!selectNextGameWord(0, -1, true)
                && !selectNextGameWord(0, -1, false)) {
            selectedCellIndex = firstGameWord?.defaultSelectionIndex ?: 0
            currentGameWord = firstGameWord
        }
    }

    /**
     * Selects next empty or errored game word depending on parameter.
     *
     * @param emptyOnly true if next empty game word should be selected, false if any errored game word shold be selected.
     */
    fun selectNextGameWord(emptyOnly: Boolean): Boolean {
        return selectNextGameWord(currentGameWord?.row ?: 0, currentGameWord?.col
                ?: 0, emptyOnly) ||
                selectNextGameWord(0, -1, emptyOnly)
    }

    /**
     * Indicates errored cells in the puzzle with a reddish background.
     *
     * @param showErrors true if errors should be indicated, false if not.
     */
    fun showErrors(showErrors: Boolean) {
        // update background of cells based on whether text is correct or not
        for (row in 0 until gridHeight) {
            for (col in 0 until gridWidth) {
                // if cell is part of currently selected game word, adjust the level to set the background color
                cellGrid[row][col]?.let {
                    val isSelected = currentGameWord == it.gameWordAcross || currentGameWord == it.gameWordDown
                    it.view?.setStyle(isSelected, showErrors && it.hasUserError)
                }
            }
        }
    }

    /**
     * @param correctly true if puzzle considered complete when everything filled in correctly, false if puzzle
     * considered complete when everthing filled in regardless of correctness.
     * @return true if puzzle is completely filled in.
     */
    fun isPuzzleComplete(correctly: Boolean): Boolean {
        for (row in 0 until gridHeight) {
            for (col in 0 until gridWidth) {
                val gridCell = cellGrid[row][col]
                if (gridCell != null) {
                    // if cell is empty, then not complete
                    if (gridCell.userChar == GameWord.BLANK || correctly && gridCell.hasUserError) {
                        return false
                    }
                }
            }
        }
        return true
    }

    /**
     * Fills in the puzzle with the user's answer for the specified game word.
     */
    fun updateUserTextInPuzzle(userText: String) {
        currentGameWord?.let {
            it.setUserText(userText.take(it.word.length))
            updateUserEntryInPuzzle(it)
            selectedCellIndex = it.defaultSelectionIndex
            showAsSelected(it, true)
        }
    }

    fun deleteLetterInPuzzle(): GameWord? {
        return updateLetterInPuzzle(GameWord.BLANK)
    }

    fun updateLetterInPuzzle(userChar: Char): GameWord? {
        currentGameWord?.let {
            it.userEntry[selectedCellIndex] = userChar
            val row = if (it.isAcross) it.row else it.row + selectedCellIndex
            val col = if (it.isAcross) it.col + selectedCellIndex else it.col
            return updateUserLetterInPuzzle(userChar, it.isAcross, row, col)
        }
        return null
    }

    fun advanceSelectedCellInPuzzle(backwardDirection: Boolean) {
        currentGameWord?.let {
            if (backwardDirection) {
                selectedCellIndex = (selectedCellIndex - 1).coerceAtLeast(0)
            } else {
                selectedCellIndex = (selectedCellIndex + 1).coerceAtMost(it.word.length - 1)
            }
            showAsSelected(it, true)
        }
    }

    //endregion

    //region PRIVATE FUNCTIONS ---------------------------------------------------------------------

    /**
     * Fills in the puzzle with the user's answer for the specified game word.
     */
    private fun updateUserEntryInPuzzle(gameWord: GameWord) {
        // show answer in puzzle
        val userEntry = gameWord.userEntry
        val wordLength = gameWord.word.length
        val isAcross = gameWord.isAcross
        var row = gameWord.row
        var col = gameWord.col
        for (charIndex in 0 until wordLength) {
            cellGrid[row][col]?.let {
                val userChar = userEntry[charIndex]
                if (isAcross) {
                    it.userCharAcross = userChar
                    col++
                } else {
                    it.userCharDown = userChar
                    row++
                }
                fillTextView(it)
            }
        }
    }

    /**
     * Fills in the puzzle with the user's letter for the specified position.
     *
     * @return word in opposing direction that had conflict and was updated if any, otherwise null
     */
    private fun updateUserLetterInPuzzle(userChar: Char, isAcross: Boolean, row: Int, col: Int): GameWord? {
        var conflictingGameWord: GameWord? = null
        cellGrid[row][col]?.let {
            if (isAcross) {
                val inConflict = it.userCharDown != GameWord.BLANK && userChar != it.userCharDown
                it.userCharAcross = userChar
                if (inConflict) {
                    it.userCharDown = userChar
                    val index = it.downIndex
                    conflictingGameWord = it.gameWordDown
                    conflictingGameWord?.let {
                        it.userEntry[index] = userChar
                    }
                }
            } else {
                val inConflict = it.userCharAcross != GameWord.BLANK && userChar != it.userCharAcross
                it.userCharDown = userChar
                if (inConflict) {
                    it.userCharAcross = userChar
                    val index = it.acrossIndex
                    conflictingGameWord = it.gameWordAcross
                    conflictingGameWord?.let {
                        it.userEntry[index] = userChar
                    }
                }
            }
            fillTextView(it)
        }
        return conflictingGameWord
    }

    /**
     * Selects next empty or errored game word depending on parameter.
     *
     * @param emptyOnly true if next empty game word should be selected, false if any errored game word should be selected.
     */
    private fun selectNextGameWord(startingRow: Int, startingCol: Int, emptyOnly: Boolean): Boolean {
        var initialCol = startingCol
        for (row in 0 until gridHeight) {
            if (row >= startingRow) {
                for (col in 0 until gridWidth) {
                    if (col >= initialCol) {
                        cellGrid[row][col]?.let {
                            var nextGameWord: GameWord? = null
                            if (row > startingRow || col > initialCol) {
                                // if the cell's word begins in the cell, then select it
                                if (it.gameWordAcross?.col == col && it.gameWordAcross != currentGameWord && isEmptyOrErroredGameWord(it.gameWordAcross, emptyOnly)) {
                                    nextGameWord = it.gameWordAcross
                                } else if (it.gameWordDown?.row == row && it.gameWordDown != currentGameWord && isEmptyOrErroredGameWord(it.gameWordDown, emptyOnly)) {
                                    nextGameWord = it.gameWordDown
                                }
                            } else if (it.gameWordDown?.row == row && it.gameWordDown != currentGameWord && isEmptyOrErroredGameWord(it.gameWordDown, emptyOnly)) {
                                nextGameWord = it.gameWordDown
                            }
                            nextGameWord?.let {
                                selectedCellIndex = it.defaultSelectionIndex
                                currentGameWord = it
                                return true
                            }
                        }
                    }
                }
                // for subsequent rows, start at first column
                initialCol = 0
            }
        }
        return false
    }

    private fun isEmptyOrErroredGameWord(gameWord: GameWord?, emptyOnly: Boolean): Boolean {
        gameWord?.let {
            return hasVisibleBlanks(it) || !emptyOnly && it.hasErroredCells
        }
        return false
    }

    private fun hasVisibleBlanks(gameWord: GameWord): Boolean {
        var row = gameWord.row
        var col = gameWord.col
        for (i in 0 until gameWord.word.length) {
            cellGrid[row][col]?.let {
                if (it.isBlank) {
                    return true
                }
            }
            if (gameWord.isAcross) {
                col++
            } else {
                row++
            }
        }
        return false
    }

    /**
     * Shows the specified word as selected (yellow highlight), or not selected.
     *
     * @param gameWord word to show as selected or unselected.
     * @param asSelected true if word should be shown as selected, false if unselected.
     */
    private fun showAsSelected(gameWord: GameWord?, asSelected: Boolean) {
        gameWord?.let {
            var row = it.row
            var col = it.col
            for (i in 0 until it.word.length) {
                if (asSelected && i == selectedCellIndex) {
                    cellGrid[row][col]?.view?.setIndividualSelection(true)
                } else {
                    cellGrid[row][col]?.view?.setSelection(asSelected)
                }
                if (it.isAcross) {
                    col++
                } else {
                    row++
                }
            }
        }
    }

    /**
     * Fills puzzle textview with the character from the user's answer.
     *
     * @param gridCell the grid cell from which to get the textview and the user's answer.
     */
    private fun fillTextView(gridCell: GridCell) {
        gridCell.view?.let {
            val text = when {
                gridCell.isConflict -> "${gridCell.userCharAcross}/${gridCell.userCharDown}"
                gridCell.isBlank -> null
                else -> gridCell.userChar.toString()
            }
            it.fillTextView(text)
        }
    }

    /**
     * Gets [GridCell] cell corresponding to the specified view.
     *
     * @param v view to get [GridCell] for.
     */
    private fun getCellForView(v: View): GridCell? {
        cellGrid.forEach {
            it.forEach {
                if (v === it?.view) {
                    return it
                }
            }
        }
        return null
    }

    //endregion

}
